Enable Union typing

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* #57013
* #57012
* #57011
* __->__ #53180

Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

Commit:
Enable Union typing

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
```
Should `optional` also be flattened here?

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (auto union_rhs = rhs_->cast<UnionType>()) {
+    return std::all_of(this->types_.begin(), this->types_.end(),
```
I think this can be slightly optimized. Right now complexity is `O(n^2)` where `n` is number of types. Since all types within `union` are sorted, I think you can search both lists of types linearly to find if one is subset of the other.

ansley replied:
I would normally just compare the two vectors, since I think `std::vector` implements `operator=`, but I did it this way because I'm worried about the subtyping relationships. What will happen if we have a class `A` and a subclass `B` and we compare `x = Union[A, ...]` with `y = Union[B, ...]`? Those `A` and `B` might not have the same index in their respective Unions, but we should still say that `y` <: `x`

gmagogsfm commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -97,6 +97,78 @@ struct SingleElementType : public Type {
+
+  // Check if a given TypePtr is an allowable member of this Union
+  bool can_hold_type(TypePtr& type) const {
+    return std::any_of(types_.begin(), types_.end(), [&](TypePtr union_contained_type){
```
Nit: 

Check if `type` is `None` and use fast path of `can_hold_none`. Same for const reference variant of this method.

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
+  if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  }
+  if (rhs_->kind() == AnyTupleType::Kind) {
```
Could you clarify why this is needed?

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union(";
```
I am a little surprised that this works since `str()` is supposed to return a legal type annotation, so in theory it should be `Union[`, what am I missing? Same for the `?` print on line 817

ansley replied:
I thought that `annotation_str` and `str` were used in different places and needed to be specified differently. Here, I'm following the example in `DictType` (for the parentheses) and `OptionalType` (the question mark representing a contained `None`).

gmagogsfm replied:
I think `annotation_str` is supposed to be valid Python type annotation, while `repr_str` is supposed to be human readable type printing used in error messages. `str` is the default implementation for both.

https://github.com/pytorch/pytorch/blob/19a0eb4cdbbe2d960c2b314352bf28b9b3cb9e9d/aten/src/ATen/core/jit_type_base.h#L74

So I think it makes sense to conform to Python's type annotation syntax, aka `Union[T1, T2, ...]`, let me know what you think.

ansley replied:
Thank you!

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -0,0 +1,345 @@
+
+    def test_unions_of_unions_are_flattened(self):
+        @torch.jit.script
+        def fn_with_union(x: Union[int, str, float]) -> str:
```
This `fn_with_union` doesn't seem necessary, as long as TorchScript accepts an `int` into `fn_with_nested_union` the test can prove union is flattened.

Same for a few other test cases where two functions are scripted, in those cases the first function doesn't seem necessary, for example `test_unions_of_a_single_argument_vanish `

Optional:
use `FileCheck` to verify union is flattened

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -0,0 +1,345 @@
+            x[1] = 2
+            return x[1]
+
+        with self.assertRaisesRegex(RuntimeError, "only int, float, "
```
Is this a fundamental limitation?

ansley replied:
No, I want to make this a follow-up issue. I found it harder to do than I expected

gmagogsfm replied:
I see, it makes sense to leave it for later. Could you file a github issue to track it? I don't think this issue is particularly urgent at all, just wanted to ensure it isn't lost.

Also, it would help if you could add `TODO` in appropriate part of code explaining that this is a temporary limitation.

ansley replied:
Issue [here](https://github.com/pytorch/pytorch/issues/58167)

gmagogsfm commented on test/test_jit.py:
```diff
@@ -2323,32 +2324,6 @@ def forward(self, input, other=four):
         t = Test()
         self.assertEqual(t(torch.ones(1)), torch.ones(1) + 4)
 
-    def test_union_to_optional(self):
```
Really glad to see these tests organized better.

gmagogsfm commented on test/tinytest.py:
```diff
@@ -0,0 +1,15 @@
+import torch
```
Should this file be in PR?

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -796,6 +796,22 @@ def is_dict(ann):
         (getattr(ann, '__origin__', None) is Dict or
             getattr(ann, '__origin__', None) is dict)
 
+def is_union(ann):
```
The commented lines can be deleted I think.

Also there is a chance to clean up redundant code inside `is_optional` to reuse `is_union`

gmagogsfm commented on torch/csrc/jit/frontend/convert_to_ssa.cpp:
```diff
@@ -93,7 +93,7 @@ struct ControlFlowLoadStores {
       auto true_type = true_vars->findInAnyFrame(x);
       auto false_type = false_vars->findInAnyFrame(x);
-      auto unified = unifyTypes(true_type, false_type);
+      auto unified = unifyTypes(true_type, false_type, true);
```
Nit:
```suggestion
      auto unified = unifyTypes(true_type, false_type, /*default_to_union*=/ true);
```

If we allow unifying to Union, then I think `unifyTypes` would never fail, which means line 97-99 wouldn't be necessary

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -994,57 +997,70 @@ struct to_ir {
+      if (!actual_return_type->isSubtypeOf(declared_return_type)) {
+        // Throw if 1) we don't have a Union, or 2) the Union does not
+        // hold the actual return type.
+        if (declared_return_type->kind() != UnionType::Kind ||
```
I am probably missing something. Isn't this condition necessary for line 1017 to be true? I think as long as line 1017 is true, we should always throw regardless of whether `union` is involved or not.

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1619,7 +1648,7 @@ struct to_ir {
       }
 
-      auto unified = unifyTypes(tv->type(), fv->type());
+      auto unified = unifyTypes(tv->type(), fv->type(), true);
```
```suggestion
      auto unified = unifyTypes(tv->type(), fv->type(), /*default_to_union=*/ true);
```

gmagogsfm reviewed (commented):
Looks great, implementation is almost bullet-proof and quite idiomatic, you must have put in a lot of thoughts in polishing it. 

I just have some minor comments plus a few questions for my own understanding.

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -85,6 +85,21 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
+        if (!union_type->can_hold_type(actual_type_ptr)) {
+          throw py::cast_error(c10::str(
+              "Expected a member of ",
+              type->expect<UnionType>()->annotation_str(),
```
```suggestion
              union_type->annotation_str(),
```

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -571,7 +571,11 @@ inline std::string friendlyTypeName(py::handle obj) {
     return ss.str();
   } else {
-    return py::str(obj.get_type().attr("__name__"));
+    auto match = tryToInferType(obj);
+    if (!match.success()) {
+      return py::str(obj.get_type().attr("__name__"));
+    }
+    return toIValue(obj, match.type()).type()->annotation_str();
```
I am probably missing something, why is this change needed? Since `Union` doesn't have a corresponding IValue tag, I am not sure how these new code relate to `Union`

ansley replied:
Reverted

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

[ghstack-poisoned]

gmagogsfm commented on test/cpp/jit/test_alias_analysis.cpp:
```diff
@@ -620,6 +620,31 @@ TEST(ContainerAliasingTest, PrimitveValuesDontAliasContainers) {
+  auto b = graph->outputs().at(1);
+  auto c = graph->outputs().at(2);
+
+  aliasDb.dump();
```
```suggestion

```

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -257,6 +292,15 @@ def fn(x: Union[List[int], int]) -> int:
 
+    def test_union_with_static_type_refinement(self):
+        def fn():
+            x: List[torch.Tensor] = []
```
Should `x` be declared to be a `Union` type?

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -334,12 +378,32 @@ def forward(self, z: str):
         self.checkModule(M(1), ("foo",))
 
+    def test_union_subtractive_refinement(self):
+        def fn(x: Union[List[int], int]):
```
This test case doesn't really test that `x` in else branch is an int because `fn` doesn't have a clear return type declared.
```suggestion
        def fn(x: Union[List[int], int]) -> int:
```

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -334,12 +378,32 @@ def forward(self, z: str):
+        self.checkScript(fn, (1,))
+        self.checkScript(fn, ([1, 2, 3],))
+
+    def test_union_subtractive_refinement_with_container(self):
```
Ditto

gmagogsfm commented on torch/csrc/jit/OVERVIEW.md:
```diff
@@ -1295,18 +1298,17 @@ and add an edge from `%output` to `%self`. The alias analysis pass is flow-insen
 Will produce a graph like this:
 
-![AliasTracker graph](https://github.com/pytorch/pytorch/blob/master/docs/source/_static/img/aliastracker_graph.png)
+![MemoryDAG graph](https://github.com/pytorch/pytorch/blob/master/docs/source/_static/img/MemoryDAG_graph.png)
```
Is this `png` checked in? I tried to access `MemoryDAG_graph.png` and it doesn't exist. But `aliastracker_graph.png` does.

gmagogsfm commented on tools/clang_format_utils.py:
```diff
@@ -130,7 +130,7 @@ def get_and_check_clang_format(verbose=False):
     # Load the reference hash and compare the actual hash to it.
     with open(hashpath, "r") as f:
-        reference_bin_hash = f.readline()
+        reference_bin_hash = f.readline().strip()
```
```suggestion
        reference_bin_hash = f.readline()
```

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -235,10 +233,47 @@ def fn2(x: Union[str, int]) -> str:
+        self.assertEqual(fn2(), 10)
+
+    def test_optional_of_union_is_flattened(self):
+        def fn() -> Union[int, str, None]:
```
Current implementation can pass when one of the following is true 1) `Optional[Union]` is flatted 2) Value binding rule allows `Optional[Union]` to be assigned to a variable of type `Union`.

since we want to make sure 1 happens, we should rule out case 2. FileCheck is a good tool to use here to verify that in IR the `Optional` type is actually flattened.

gmagogsfm commented on test/cpp/jit/test_alias_analysis.cpp:
```diff
@@ -620,6 +620,31 @@ TEST(ContainerAliasingTest, PrimitveValuesDontAliasContainers) {
   }
 }
 
+TEST(ContainerAliasingTest, UnionAliasing) {
```
Could you test the case that Elias provided as well? IIRC, it is about `List[Union[Tensor, int]]`

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
+    return a->kind() > b->kind();
+  });
+  // Filter out duplicate types
+  types_.erase(std::unique(types_.begin(), types_.end()), types_.end());
```
Are you sure std::unique works here ? Won't that that be comparing pointer equality ? I know that there is a unique pointer for like Bool, but I'm not sure that guarantee exists for Lists/Dicts etc

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
```
this will only guarantee stable sorting for types with the same kind, but e.g Union[List[int], List[float]] types will not equal Union[List[float], List[int]]. I think we're relying on the ordering in `==` and maybe elsewhere

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
+}
+
+  // Check if a given TypePtr is an allowable member of this Union
+  bool UnionType::can_hold_type(TypePtr& type) const {
```
Why do we need const and non const version ?

ansley replied:
There are lots of places where the `const`-ness is already set in stone. I didn't think it made sense to trace back all those calls and change them all to be non-`const` so that they could work with Union, since that could have implications for the larger codebase. I might have underestimated the importance of keeping the API small, though. What would you suggest?

eellison replied:
I guess i figured we just needed a const version, but it's not a big deal

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
+    if (type == NoneType::get()) {
+      return can_hold_none_;
+    }
+    return std::any_of(types_.begin(), types_.end(), [&](TypePtr union_contained_type){
```
How does this interact with other unions/optionals

ansley replied:
`isSubtypeOf` handles the subtyping relationship between Unions and Optionals. I consider `Union[T1, T2]` to be a subtype of `Union[T1, T2, T3]`, for example.

eellison reviewed (commented):
Part of the way through review. I think Optionals should Inherit from UnionType, and we should always return Optionals when we detect them in Union instant oat ion. As is, I think there a lot of funny cases with Union[None, T] that will interact weirdly with OptionalType. One other thing is that I saw that serialization isn't implemented yet. That's going to be not the nicest user failure. Could you stack the serialization PR on top of this one and land both at the same time ?

ansley commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1541,6 +1607,12 @@ TORCH_API std::ostream& operator<<(std::ostream& os, const Stride& s);
 // what is the type, ignoring extra size/shape information?
 // e.g. Tensor(2x3) -> Dynamic, and Tuple(Tensor(2x3),...) -> Tuple(Dynamic,...)
 
+// `unshapedType` is used to remove Tensor subtypes to provide more
```
What else is `unshapedType` used for? I had a hard time understanding it just by looking around the codebase, so I want to add some sort of documentation

eellison replied:
if you search through code base there are 14 uses. https://github.com/pytorch/pytorch/search?p=2&q=unshapedType  . It's useful anytime you want to refine tensor types and then unrefine it.

gmagogsfm commented:
> Part of the way through review. I think Optionals should Inherit from UnionType, and we should always return Optionals when we detect them in Union instant oat ion. As is, I think there a lot of funny cases with Union[None, T] that will interact weirdly with OptionalType. One other thing is that I saw that serialization isn't implemented yet. That's going to be not the nicest user failure. Could you stack the serialization PR on top of this one and land both at the same time ?

Good point, `optional` should ideally inherit from `union`. Could we do it as a next step though? I think this PR is quite large, and we have quite a few users blocked on lack of `union`. Landing this PR would unblock them.

As for serialization, I think it is supported. The only missing piece is from unpickler. However `union` doesn't need to be handled by unpickler at all, since there is never going to be an IValue of `union` type.

eellison commented:
@gmagogsfm I'm find delaying the Optional inheriting from Union if `Optional[int] == Union[None, int]` doesn't work that's a bug and I think this is the easiest to fix it.

gmagogsfm commented:
> @gmagogsfm I'm find delaying the Optional inheriting from Union but if `Optional[int] == Union[None, int]` doesn't work that's a bug and I think this is the easiest to fix it.

Yep, agreed that is more principled way of implementing it. I think `optional[t] == Union[None, T]` currently works, see `test_union.py:235`.

eellison commented:
@gmagogsfm i'm pretty sure that's testing Subtyping not `==`.

eellison commented:
@gmagogsfm if you instantiate a List[Union[int, None, float]] you will have an IValue with a Union type, I think it does need to be handled

ansley commented:
@eellison This test passes:

```python
@torch.jit.script
def foo(x: List[Union[int, None]]):
    pass

@torch.jit.script
def fi(x: List[Optional[int]]):
    return foo(x)

@torch.jit.script
def foo2(x: List[Union[int, None]]):
    return fi(x)
```

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -752,46 +752,90 @@ TupleTypePtr TupleType::createNamed(
-    return a->kind() > b->kind();
+  std::sort(types.begin(), types.end(),
+            [](const TypePtr a, const TypePtr b) -> bool {
+              return a->str() > b->str();
```
Curious why str comp is used instead

ansley replied:
Elias cleverly pointed out that my previous comparison [would not guarantee stable sorting](https://github.com/pytorch/pytorch/pull/53180#discussion_r610957012). I don't need the `TypePtr`s sorted in any special order; I just need to guarantee that a given set of `TypePtr`s will always be sorted in the same order. So, sort by their serialized representation.

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -305,6 +308,8 @@ def try_ann_to_type(ann, loc):
         msg = "Unsupported annotation {} could not be resolved because {} could not be resolved."
         assert valid_type, msg.format(repr(ann), repr(contained))
         return OptionalType(valid_type)
+    if is_union(ann):
```
Just making sure I understand the plan, we want to allow users to directly annotate values with `Union` from day 1?

gmagogsfm replied:
What is the plan here? It is a bit dangerous to enable it from day 1 because of FC. production services may not be updated to parse Union yet. But our serialized model from newer PyTorch versions would contain them, thus causing production crashes.

gmagogsfm reviewed (approved):
Looks good to me, just left some questions about our plan of deployment.

@eellison Could you take a look as well? This is a blocker for PDT, which we are hoping to get out by 1.9. Thanks in advance!

ansley commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -97,52 +97,87 @@ struct SingleElementType : public Type {
+// This represents any type that we can dynamically specialize at
+// runtime. Currently only `Optional` and `Union` inherit from
+// `DynamicType`
+struct DynamicType : public Type {
```
`Optional` used to inherit from `SingleElementType`, which exposed certain methods that I felt made sense in `Optional` but didn't make sense in `Union`. Also, `Union` has certain methods that didn't make sense in `Optional` as Optional is CURRENTLY used in the codebase. 

I have plans to change this later, though. It will be possible in my follow-up project where I remove all internal instances of `Optional` and make `Optional[T]` a simple alias for `Union[T, None]` [the way Python does](https://github.com/python/typing/blob/master/src/typing.py#L850).

ansley replied:
Thanks for the feedback, changing now!

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3547,6 +3556,99 @@ struct to_ir {
+    auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
+
+    // determine the element type of the list
+    // if we have a type hint of List[T], use T
+    // if the list is non-empty use type_of(list[0])
+    // otherwise assume it is List[Tensor]
+    TypePtr elem_type = TensorType::get();
+    if (type_hint) {
+      if (type_hint->kind() == TypeKind::ListType) {
+        elem_type = type_hint->expectRef<ListType>().getElementType();
+      } else {
+        // If the type hint was not a List[T] throw an error
+        throw ErrorReport(ll) << "Expected a List type hint but instead got "
+                              << type_hint->repr_str();
+      }
+    } else if (!values.empty()) {
+      std::stringstream ss;
+      auto types = fmap(values, [](const Value* v) { return v->type(); });
+      auto maybe_elem_type = unifyTypeList(types, ss);
+      if (!maybe_elem_type) {
+        throw ErrorReport(ll) << "Lists must contain only a single type\n"
+                              << ss.str();
+      }
+      elem_type = maybe_elem_type.value();
+    }
+
+    for (auto v : values) {
+      std::stringstream ss;
+      if (!v->type()->isSubtypeOfExt(elem_type, &ss)) {
+        throw ErrorReport(ll)
+            << "Lists must contain only a single type, expected: "
+            << elem_type->repr_str() << " but found " << v->type()->repr_str()
+            << " instead.\n"
+            << ss.str();
+      }
+    }
+    Value* result =
+        graph->insertNode(graph->createList(elem_type, values))->output();
+    return result;
+  }
```
I had originally started to modify it, since I found a test case in which creating a `ListLiteral` with `Union` or `Optional` fails:

```python
@torch.jit.script
def fn():
    x: Union[List[torch.Tensor], int] = [torch.tensor(3)]
    if isinstance(x, List[torch.Tensor]):
        x.append(torch.tensor(3))
    return x
```

The problem is that the `ListConstruct` is assigned to a Value of type `Union[int, Tensor[]]`. That makes sense because of the `isinstance` check later--if we didn't assign it to be `Union` type, the implied else branch where `x` is an `int` would be an error. However, that fails in the interpreter because the `listConstruct` instruction (interpreter version, note the casing) tries to cast the given element to a `List` when it's typed as a `Union`.

Because we need this out ASAP, I decided to put back the original code and just remove the failing test from the `test_union.py` suite. But I thought it looked a lot better to factor this logic out into a helper function, which is why I kept it this way.

eellison replied:
If the code is not modified, please leave the refactor in a separate ghstack, it makes it hard to review as is.

eellison commented on torch/csrc/jit/OVERVIEW.md:
```diff
@@ -1295,18 +1298,17 @@ and add an edge from `%output` to `%self`. The alias analysis pass is flow-insen
-  else:
-    r = b[0]
-  return c, r
+def foo(a: Tensor, b: Tensor):
```
Also good changes, but IMO would be better in a separate PR given length of this one

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1156,6 +1175,19 @@ struct to_ir {
     }
+    if (const auto& union_type = lhs_value->type()->cast<UnionType>()) {
+      std::vector<Refinement> present;
+      for (const auto& type : union_type->types()) {
```
You're refining the name to each contained type that isn't None. What we want is to refine the type to its contained types minus NoneType.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1700,7 +1733,15 @@ struct to_ir {
           return true;
         }
-        if (auto op = actual_type->cast<OptionalType>()) {
+        if (const auto& op = actual_type->cast<UnionType>()) {
```
When Optional inherits from Union, can we refactor this and the Optional handling ?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1724,16 +1765,38 @@ struct to_ir {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+      Refinement isinstance(ident, gathered.types.at(0));
```
Can you factor out the refinement logic into a separate stacked PR? I am having a really hard time parsing through the logic and the 500 lines to verify that this is properly being tested.  There was the previous bug I commented with refining Unions even though @gmagogsfm which is a good indication that this PR is too large for us to be confident about code quality

eellison reviewed (commented):
Have not yet done 
- alias analysis
- type refinement


We discussed offline about some other changes. Could you please separate out the alias analysis and type refinement parts of the PR into separate ghstacks with corresponding tests on top of this PR ?

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+    contained_ = types[0]->kind()!= NoneType::Kind ? types[0] : types[1];
+  }
+  else {
+    contained_ = UnionType::create(types);
```
having a Optional[Union] seems wrong...

ansley replied:
This goes back to the problem we had with the `Union` constructor and factory function earlier...if I'm stuck returning `OptionalType`, what's the best thing to do when I actually want to return `Union[T1, ..., TN, None]`? Throw an error?

eellison reviewed (commented):
Hmm, gave a review, but i could really use someone to double check intersection/set difference. Is that in the context of type refinement? maybe cc @bhosmer @gmagogsfm 

The other thing is that having an Optional[Union] is not great but I'm not sure what the best way of getting rid of it is.

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+              types.end());
+}
+
+UnionType::UnionType(std::vector<TypePtr> types, TypeKind kind) : Type(kind) {
```
Good idea. Let's think about other possible solutions before this ships 
cc @gmagogsfm

ansley replied:
One thing I wanted to explicitly point out to make sure we're all on the same page...the only way that an empty `Union` or a `Union` with one type can be created is using C++. If a user goes through `torch.jit.script`, the Python interpreter will throw for an empty `Union` and automatically use the internal type a single-type `Union`. (In the latter case, the `Union` constructor actually [returns the internal type](https://docs.python.org/3/library/typing.html#typing.Union).) So, the natural flow of our code should prevent most user error

bhosmer commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  return union_of(rhs->types());
+}
+
+std::vector<TypePtr> UnionType::intersection_of(c10::ArrayRef<TypePtr> rhs_types) const {
```
Yeah - if there's a subtype relation between types, then you need to include subtypes of subtype/supertype pairs when taking an intersection. 

Note that you need to make sure that the UnionType you're intersecting against is canonicalized to not have any internal overlap among its types - members shouldn't be subtypes of each other. Otherwise the results of an intersection can become order-dependent.

You can do the canonicalization at construction time and in `union_of` by being sensitive to subtypes in the opposite way - including supertypes of subtype/supertype pairs, and omitting/removing the subtypes.

ansley replied:
Good point. I've updated the algorithms, so please let me know if they look good when I push the amended version

bhosmer commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+
+std::vector<TypePtr> UnionType::set_difference_of(c10::ArrayRef<TypePtr> rhs_types) const {
+  std::vector<TypePtr> types;
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{rhs_types.begin(), rhs_types.end()};
```
I don't actually think the system is closed under this kind of directional difference - like say `types` has `Animal` and `rhs_types` has `Cat`, with `Cat <: Animal`. You'd need a way to express `Animal - Cat` as a type, i.e. "supertype except for subtype".

Where would this be used? There might be coarse-grained things you could do instead, or compatibility conditions you could impose, but it'd depend on the use case whether any of those moves were reasonable...

bhosmer replied:
Yeah, I'd say ditch difference but intersection and union are well behaved and will definitely come in handy during type propagation, subtype testing of unions against each other, etc.

ansley replied:
Cool, thanks

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (rhs_->kind() == NoneType::Kind && can_hold_none_) {
```
You're totally right! Thank you for thinking through this

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  if (rhs_->kind() == NoneType::Kind && can_hold_none_) {
+    return true;
+  }
+  else if (const auto union_rhs = rhs_->cast<UnionType>()) {
```
Thanks!

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  else if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors
+    if (this->types() == union_rhs->types()) {
```
Changed to `TypeEqual`

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  types.reserve(this->types().size() + rhs_types.size());
+  types = this->types().vec();
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{types.begin(), types.end()};
+  std::copy_if(rhs_types.begin(), rhs_types.end(),
```
Sounds good, I'll fix it. 

I originally wanted to have this return a `std::vector` for greater flexibility in the codebase. [This](https://github.com/pytorch/pytorch/pull/57011/files#diff-0583c36ed597bbc23769918d2e3086c98bf2dfd7a79acdb623a31c6f1aa787c8R1180) is an example of where returning a `std::vector` is easier to work with (and read later) than constructing a new `UnionType` only to dismantle it again for its internal vector.

On the other hand, we'd want to coalesce types where we can the way we do in the `Union` constructor. So I think it makes the most sense to just create a new Union like you're saying.

Leaving this comment open for discussion later, though @gmagogsfm would probably agree that these methods should return a new `Union` as well...

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -748,6 +755,247 @@ TupleTypePtr TupleType::createNamed(
+  return intersection_of(rhs->types());
+}
+
+bool UnionType::is_optional_of_type(const TypePtr type) const {
```
Deleted the method, thanks

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+    });
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
```
There are probably some cases where there are more than 2 types in `types_` while subtyping relationship still holds. Think of following case:

```
lhs: Union[T_a, T_b, None]
rhs: Optional[T_c]

Subtyping relationship:
T_c > T_a
T_c > T_b
T_a has no relationship with T_b

```

Then lhs is a subtype of rhs, but this check would fail

eellison replied:
lhs is not a subtype of rhs in this example.

gmagogsfm replied:
Could you clarify? All types in `lhs` are subtypes of `Optional[T_c]` in this example.

eellison replied:
Yea, nevermind, misread the example.

ansley replied:
Great point, Yanan

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  // decrement `end`, swap `types[j]` with the unified type, and
+  // break. Otherwise, we keep `end` where it is to signify that the
+  // new end of the vector hasn't shifted
+  size_t end_idx = types.size()-1;
```
If T_a is a supertype of T_b, and T_b is a supertype of of T_c, T_a is necessarily a supertype of T_c

gmagogsfm replied:
Yeah makes sense, I was mainly worried about enforcement, our subtype checking can be quite free-form, wasn't sure if any invariant is kept.

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+}
+
+void standardizeUnion(std::vector<TypePtr>& types) {
+  filterDuplicateSubtypes(types);
```
I think we need to filterDuplicateSubtypes and then flatten, instead of the opposite, which we're doing now. Otherwise we'll end up with Union[Optional[int], Optional[float]] instead of Union[int, float, None]

ansley replied:
I'm not sure what you mean...if we have `Union[Optional[int], Optional[float]]`, `flattenUnion` will make it `Union[int, float, None]`, and then `filterDuplicateSubtypes` won't have any work to do, right? So `flattenUnion` should come first

ansley replied:
Is this good?

```python
    def test_union_redundant_arguments_are_skipped_optional(self):
        @torch.jit.script
        def fn(x: Union[int, Optional[float], Optional[int]]) -> str:
            return "foo"

        s = fn.graph

        FileCheck().check("x : Union[float, int, NoneType]")    \
                   .run(s)
```

eellison replied:
i think this is relying on python's filtering of duplicate subtypes and unification, not our own

ansley replied:
Thought I could get out of writing C++ tests ;(

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  // Coalesce types and delete all duplicates. Moving from right to left
+  // through the vector, we try to unify the current element (`i`) with
+  // each element (`j`) before the "new" end of the vector (`end`).
+  // If we're able to unify the types at `types[i]` and `types[j]`, we
```
this is describing the code as it exists, could you try to explain the motivation behind it?

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  if (const auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors. We need to use
+    // `std::equal` with `TypeEqual` for optimal performance; do not
```
We're bypassing the `==` handling of Union/Optionals here because we are checking kind

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  return UnionTypePtr(std::move(union_type));
+}
+
+bool UnionType::operator==(const Type& rhs) const {
```
Shouldn't we add the equivalent union/optional `==` logic in `OptionalType::operator==` ?

ansley replied:
Do we not have it in there already? It's in `jit_type.h`

eellison replied:
Not for equality with Unions

eellison replied:
can you add a test for this ?

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
+      (types_[0]->isSubtypeOf(inner) ^ types_[1]->isSubtypeOf(inner));
```
nit: don't use `^`

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  } else if (const auto optional_rhs = rhs_->cast<OptionalType>()) {
+    TypePtr inner = optional_rhs->getElementType();
+    return can_hold_none_ && types_.size() == 2 &&
+      (types_[0]->isSubtypeOf(inner) ^ types_[1]->isSubtypeOf(inner));
```
Can `Optional` can't have an `inner` type that None subtypes? I think that would be a malformed type. Still i think this could be more clear to follow with `(types_[0] != Nonetype::get() && types_[0]->isSubtypeOf(inner)) || (types_[1] != Nonetype::get() && types_[1]->isSubtypeOf(inner))`

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+
+UnionTypePtr UnionType::intersection_of(const UnionTypePtr rhs) const {
+  std::vector<TypePtr> types;
+  std::unordered_set<TypePtr, std::hash<TypePtr>, TypeEqual> dict{rhs->types().begin(), rhs->types().end()};
```
im a little nervous about the dict caching logic. what is it trying to achieve ?

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+    }
+    // If we can find some common supertype with a type in `rhs`
+    for (const TypePtr rhs_type : rhs->types()) {
+      if (auto unified = unifyTypes(lhs_type, rhs_type)) {
```
we shouldn't be using `unifyTypes` here, for example intersection of `None`, `int` is an empty set, but unifyTypes will yield `Optional[int]`. Could we add some tests to this function ?

eellison reviewed (commented):
Nice, looking good! another round of comments

gmagogsfm reviewed (changes requested):
Undoing my previous approval since there has been a lot of changes in code

bhosmer commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+    for (size_t j = std::min(i-1, end_idx); ; --j) {
+      c10::optional<TypePtr> unified = wrapper(types[i], types[j]);
+      if (unified) {
+        types[j] = *unified;
```
I don't think unifying is what you're after here - you just need to test for subtyping. Unification will find a common supertype even if the two args aren't directly in a subtype relationship, so in those cases you'll be merging rather than just deduping, which you don't want.

Like if A > B and A > C, and you have Union[B, C], with unification you'll wind up with Union[A], but that's a different type - you don't know what else A has in it besides B and C.

OTOH if you have Union[A, C], then you do want to wind up with Union[A], bc C is completely contained in A, so the two unions are equivalent in terms of the values they represent.

ansley replied:
cc @gmagogsfm @eellison 
I agree with Basil's point, and I'm changing the code accordingly. Basil and I talked offline, and we discussed that, right now, `unifyTypes` won't have the `A, B, C` problem because of how limited our type system is. So, this change is for forward compatibility more than anything. I wanted to flag this for the next round of reviews

bhosmer reviewed (commented):
Hi @ansley - lots of good work here, thanks for pinging me to take another look! LMK if any of the inline comments aren't clear - I think there are a few simplifications that will make life easier. (BTW going by the titles, you might be doing something like my OptionalType suggestion further up the stack, in which case you might wanna think about resequencing, as I think a lot of the OptionalType-specific code here will become unnecessary.)

ansley commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -93,25 +93,107 @@ struct SingleElementType : public Type {
+
+  bool operator==(const Type& rhs) const override;
+
+  at::ArrayRef<TypePtr> types() const {
```
Thanks!

ansley commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -93,25 +93,107 @@ struct SingleElementType : public Type {
+
+  friend struct Type;
+
+  bool operator==(const Type& rhs) const override {
```
Thanks! I'm going to leave these comments up and address them in #58673

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -343,7 +343,6 @@ c10::optional<TypePtr> unifyTypeList(
     why_not << "Cannot get unified type from empty list";
     return c10::nullopt;
   }
-
```
Thanks, won't happen again

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+      flattenUnion(inner_type, types);
+    }
+  } else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
```
Thanks

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+
+// Remove nested Optionals/Unions during the instantiation of a Union or
+// an Optional. This populates `types` with all the types found during
+// flattening
```
Amended comment, thanks

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  TORCH_INTERNAL_ASSERT(types_.size() != 1, "After type unification was"
+                        " performed, the Union only has one type. "
+                        "Use the common supertype instead of creating "
+                        "a Union type");
```
Thanks

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+                        "Use the common supertype instead of creating "
+                        "a Union type");
+
+  TORCH_INTERNAL_ASSERT(types_.size() >= 2, "Cannot create a Union of "
```
Thanks

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+  return intersection_of({rhs_types});
+}
+
+UnionTypePtr UnionType::intersection_of(const UnionTypePtr rhs) const {
```
I talked with @gmagogsfm offline about this, and we agreed that it would be best to delete `union_of` and `intersection_of` (sadly). If we're all arguing about what "intersection" means when you take subtyping relationships into account, then I don't trust that future users of this API will be able to tell what it means, either.

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+    // If we can find some common supertype with a type in `rhs`
+    for (const TypePtr rhs_type : rhs->types()) {
+      if (auto unified = unifyTypes(lhs_type, rhs_type)) {
+        types.emplace_back(*unified);
```
Thanks! Per our discussion, I'm just deleting these helper methods

ansley commented on test/cpp/jit/test_union.cpp:
```diff
@@ -0,0 +1,46 @@
+    ASSERT_TRUE(opt->isSubtypeOf(union2));
+    ASSERT_FALSE(opt->isSubtypeOf(union3));
+
+    ASSERT_FALSE(union1->isSubtypeOf(none));
```
Good call

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -649,7 +652,8 @@ struct to_ir {
           << "methods must have a self argument";
     }
-    method.setSchema(emitDef(def, self, graph->block()));
+    auto schema = emitDef(def, self, graph->block());
```
Oops, stray debugging change

ansley commented on torch/csrc/jit/frontend/schema_type_parser.cpp:
```diff
@@ -331,6 +332,18 @@ std::pair<TypePtr, c10::optional<AliasInfo>> SchemaTypeParser::parseType() {
     L.expect(')');
     alias_info = parseAliasAnnotation();
     value = DictType::create(key_type, value_type);
+  } else if (L.cur().kind == TK_IDENT && L.cur().text() == "Union") {
```
Actually, does it need to be in there in the first place? We shouldn't see `Union` in a schema ever

eellison replied:
why not ?

ansley replied:
We don't allow the keyword `Union` in a schema yet. If we have `Union[T, None]`, it's represented as `?` for backwards compatibility purposes

eellison replied:
Sure, if we want to leave it as unimplemented that's fine, we should delete the parsing logic in that case then instead of leaving it in untested.

ansley commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -93,25 +93,107 @@ struct SingleElementType : public Type {
+    return has_free_variables_;
+  }
+
+  UnionTypePtr union_of(std::vector<TypePtr>& rhs_types) const;
```
Deleting method per earlier conversation

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

eellison commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -93,25 +93,112 @@ struct SingleElementType : public Type {
+  // an Optional (in which case only `containedTypes` is overridden for
+  // BC purposes). This method should be deleted once we canonicalize
+  // `Optional[T]` as `Union[T, None]`
+  at::ArrayRef<TypePtr> getTypes() const {
```
i don't know why we can't use `containedTypes` here

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <c10/util/irange.h>
 #include <ATen/core/grad_mode.h>
 #include <ATen/core/function.h>
+#include <torch/csrc/jit/ir/type_hashing.h>
```
i think this include isnt needed

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -753,6 +764,265 @@ TupleTypePtr TupleType::createNamed(
+  if (rhs->kind() == OptionalType::Kind) {
+    return true;
+  }
+  if (auto rhs_ = rhs->cast<UnionType>()) {
```
do wee need to add this logic? won't it naturally be covered by `Type::isSubtypeOfExt` ?

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented on torch/jit/annotations.py:
```diff
@@ -323,6 +327,8 @@ def try_ann_to_type(ann, loc):
         assert valid_type, msg.format(repr(ann), repr(contained))
         return OptionalType(valid_type)
+    if is_union(ann):
+        return UnionType([try_ann_to_type(a, loc) for a in ann.__args__])
```
By "the other stuff here", do you mean my logic? Or do you mean the other types that use `[try_ann_to_type(a, loc) for a in ann.__args__]`?

ansley commented on test/test_jit.py:
```diff
@@ -13087,8 +13062,8 @@ def fn(x):
         self.checkScript(fn, ("y"))
 
         def index_str_to_tensor(s):
-            # type: (str) -> int
```
My code puts stricter requirements on type annotations matching the actual return type, so this test fails due to the incorrect type annotation. I think that the `-> int` was just a typo due to 1) the name of the test (`index_str_to_tensor`) and 2) the fact that we're just calling `checkScript` on it (i.e. it doesn't look like the author expected it to fail).

ansley replied:
I added stricter type annotation requirements in IR Emitter, which causes this test to fail. The test passes if I change the annotation from `-> int` to `Tensor`. I think that the original annotation was just a mistake for two reasons: 1) the name of the test is `index_str_to_tensor`, which makes it sound like the return type should be a Tensor, and 2) the test is only checked using `checkScript` instead of something that would imply an error is expected (e.g. `assertRaisesRegex`)

bhosmer commented on torch/nn/functional.py:
```diff
@@ -4716,15 +4716,17 @@ def _in_projection_packed(
     if k is v:
         if q is k:
             # self-attention
-            return linear(q, w, b).chunk(3, dim=-1)
```
@ansley per offline chat, it'd be better to fix your test case by using `# type: ignore[return-value]` here - since `chunk()` actually does return a tuple, the new code on line 4720 is literally just building a second tuple from the members of the first. 

Similarly on line 4727, though you might need `# type: ignore[operator, return-value]` there. (LMK if this doesn't fix the errors you were seeing - we might need to do something more, but one way or another I'm sure there's a way to quiet mypy without doing the gratuitous tuple-rebuilding at runtime.)

Also though, I'd suggest breaking this out to a separate PR, since this isn't really thematically related to the rest of what you're up to here.

cc @gmagogsfm - background here is that `torch.chunk()` has a schema return type of `Tensor[]`, but actually returns a Python `Tuple[Tensor, ...]` in eager.

ansley replied:
Fixed. See Basil's comment

bhosmer reviewed (changes requested):
Hi @ansley - the change request is just re the type annotations in `functional.py`, nothing to do with the union typing work.

ansley commented on aten/src/ATen/core/jit_type_base.h:
```diff
@@ -17,6 +17,7 @@ namespace c10 {
   _(TensorType)             \
   _(StorageType)            \
   _(TupleType)              \
+  _(UnionType)              \
```
Fixed

ansley replied:
Thanks for telling me about this!

ansley commented on torch/_jit_internal.py:
```diff
@@ -823,33 +823,28 @@ def is_dict(ann) -> bool:
-        if len(args) == 2:
-            union_optional = (safe_is_subclass(args[1], type(None)) and not safe_is_subclass(args[0], type(None))) \
-                or (safe_is_subclass(args[0], type(None)) and not safe_is_subclass(args[1], type(None)))
+    def _is_optional(ann):
```
Fixed

ansley replied:
Fixed

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -557,8 +559,16 @@ bool Type::isSubtypeOfExt(const TypePtr& rhs, std::ostream* why_not) const {
+  }
+  if (auto union_rhs = rhs->cast<UnionType>()) {
+    // Check if `this` is a subtype of any of the types within the Union
+    return std::any_of(union_rhs->containedTypes().begin(),
```
I can't call `canHoldType` on `this`

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+
+  standardizeUnion(types_);
+
+  TORCH_INTERNAL_ASSERT(types_.size() != 1, "After type unification was"
```
I agree, but it's harder because this is the constructor (so, we inherently have to return a `UnionType`). I feel strongly that we should NOT allow "Unions" of a single type, since we would have to add in a ton of extra checks to ensure that the semantic remain the same

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+    // `union_rhs->types_`
+    return std::all_of(this->types_.begin(), this->types_.end(),
+                       [&](TypePtr t) {
+                         return union_rhs->canHoldType(t);
```
Good catch, fixed

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+    }
+    auto optional_lhs = this->toOptional();
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
```
Thanks!

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+    if (optional_lhs) {
+      return *optional_rhs == *((optional_lhs.value())->cast<OptionalType>());
+    }
+    return false;
```
Fixed

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+  }
+  filterDuplicateSubtypes(test_types);
+
+  auto get_index_if_exists = [&](TypePtr lhs) -> c10::optional<decltype (test_types.begin())> {
```
Yeah, I just tested it, and it looks like we need one of the `decltype`s. Compiler error:

```
../aten/src/ATen/core/type.cpp: In lambda function:
../aten/src/ATen/core/type.cpp:1035:17: error: inconsistent types ‘c10::optional<__gnu_cxx::__normal_iterator<std::shared_ptr<c10::Type>*, std::vector<std::shared_ptr<c10::Type> > > >’ and ‘c10::nullopt_t’ deduced for lambda return type
     return c10::nullopt;
                 ^~~~~~~
ninja: build stopped: subcommand failed.
```

I can take out the other one, though!

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+    return c10::nullopt;
+  };
+
+  auto int_idx = get_index_if_exists(IntType::get());
```
TODO when we have more types in the type system. Right now, it's not worth the drop in readability

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+              });
+
+  if (types.size() == 0) {
+    return c10::optional<TypePtr>{};
```
Thanks!

ansley commented on aten/src/ATen/core/type.cpp:
```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+      return getElementType()->isSubtypeOfExt(optional_rhs->getElementType(), why_not);
+    }
+    if (UnionTypePtr union_rhs = rhs->cast<UnionType>()) {
+      return union_rhs->canHoldType(NoneType::get()) && union_rhs->canHoldType(this->getElementType());
```
Fixed

ansley commented on torch/_jit_internal.py:
```diff
@@ -823,33 +823,28 @@ def is_dict(ann) -> bool:
+    if ann is Optional:
+        raise_error_container_parameter_missing("Optional")
 
-    union_optional = False
```
I've looked at the code as well as the blame on Phabricator, and I'm not sure how this could break BC. To me, it seems as though it's a way to see if we have `Union[T, None`, which can be parsed as `Optional`. Could you please let me know which Python version this was put in here for?

driazati replied:
afaik `Union[T, None]` and `Optional[T]` have always been equivalent to Python, but at some point they switched from desugaring to `Union` to `Optional`:

python 3.6

```python
>>> import typing
>>> print(typing.Union[int, None])
typing.Union[int, NoneType]
>>> print(typing.Optional[int])
typing.Union[int, NoneType]
>>> print(typing.Union[int, None] == typing.Optional[int])
True
```

python 3.8

```python
>>> import typing
>>> print(typing.Union[int, None])
typing.Optional[int]
>>> print(typing.Optional[int])
typing.Optional[int]
>>> print(typing.Union[int, None] == typing.Optional[int])
True
```

ansley replied:
Now that we've introduced Union, we should be fine whether we have 3.6 or 3.8, though

ansley commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1009,57 +1012,66 @@ struct to_ir {
+    }
+    AT_ASSERT(declared_return_type);
+
+    if (declared_return_type->kind() == UnionType::Kind) {
```
Deleted

ansley commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -65,6 +65,12 @@ TypePtr ScriptTypeParser::subscriptToType(
     return OptionalType::create(elem_type);
 
+  } else if (typeName == "Union") {
+    std::vector<TypePtr> subscript_expr_types;
```
Fixed

ansley commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -977,7 +979,7 @@ bool Node::mustBeNone() const {
       (outputs().size() == 1 && output()->type() == NoneType::get()) ||
       // It's a constant optional with no value in the attributes.
       (kind_ == prim::Constant && !this->hasAttributes() &&
-       output()->type()->cast<OptionalType>());
```
Reverted

ansley commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1694,20 +1697,23 @@ Node* Graph::createEnumValue(Value* e) {
   return n;
 }
 
-Node* Graph::createList(const TypePtr& elem_type, at::ArrayRef<Value*> values) {
```
Thanks for the reminder. I'm not making any more of these changes in newer versions of this PR. These are leftovers from when I didn't know that small changes like this were bad practice

ansley commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -43,7 +43,10 @@ class TypeParser {
       return CreateSingleElementType<ListType>();
     } else if (token == "Optional") {
-      return CreateSingleElementType<OptionalType>();
+      expect("[");
```
Fixed

ansley commented on torch/csrc/jit/python/python_ir.cpp:
```diff
@@ -864,6 +864,16 @@ void initPythonIRBindings(PyObject* module_) {
+        for (const auto& type : self.containedTypes()) {
+          types.push_back(type);
+        }
+        return types;
```
Fixed

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -909,22 +924,27 @@ bool UnionType::operator==(const Type& rhs) const {
-             && this->canHoldType(NoneType::get())
-             && this->canHoldType(NumberType::get());
+             && this->can_hold_none_
+             && can_hold_numbertype();
```
TODO:
Maybe this can be made more generic by calling `can_hold_numbertype` inside `canHoldType`

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -909,22 +924,27 @@ bool UnionType::operator==(const Type& rhs) const {
+      return *optional_rhs == *((optional_lhs.value())->expect<OptionalType>());
+    } else {
+      return false;
     }
```
```suggestion
    }
```

gmagogsfm commented on test/test_jit.py:
```diff
@@ -6253,23 +6253,24 @@ def func(a, b):
-        for module in [M1, M2, M3]:
-            self.checkModule(module().train(), ())
-            self.checkModule(module().eval(), ())
+    # Uncommented in next PR
```
`@skipIf`

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -331,7 +330,17 @@ def try_ann_to_type(ann, loc):
+            if a is None:
+                inner.append(NoneType.get())
+            maybe_type = try_ann_to_type(a, loc)
+            if maybe_type is not None:
```
We should error out if `maybe_type` is `None`

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

gmagogsfm commented on test/expect/TestScript.test_annot_ast_mypy_fn.expect:
```diff
@@ -6,4 +6,4 @@ foo(bool x, (Tensor, Tensor) y) -> ((bool, bool))
 foo(float[3] x, (Tensor, Tensor) y) -> ((float[], float[]))
```
Are these `expect` file changes intended? Looks like it is adding empty line at end of each file?

garymm commented:
@ansley any blockers to merging this?

bhosmer reviewed (approved):
Looks like the thing I was requesting changes to is long since fixed, accepting to unblock 😁

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26779086).

Commit:
Update on "Enable Union typing"


Differential Revision: [D26779086](https://our.internmc.facebook.com/intern/diff/D26779086)

fixes #41412

[ghstack-poisoned]

garymm commented:
@ansley it'd be really nice to have this in. Are there blockers?

garymm commented:
Yay thanks for this!

Thank you for the kind words on the documentation PRs. I will certainly ask for your review in the future.

ProGamerGov commented:
Someone should update the JIT page that says `Union` is unlikely to be supported: https://pytorch.org/docs/stable/jit_language_reference.html#language-reference

Looks like the change to fix it should be in either: https://github.com/pytorch/pytorch/blob/master/docs/source/jit_language_reference.rst or https://github.com/pytorch/pytorch/blob/master/docs/source/jit_language_reference_v2.rst

gmagogsfm commented:
Thanks! V1 langref is already updated, just made a PR to update lang ref v2 #69514
